go.mod

```diff
@@ -0,0 +1,136 @@
+	gopkg.in/go-playground/assert.v1 v1.2.1 // indirect
+	gopkg.in/inf.v0 v0.9.1 // indirect
+	gopkg.in/natefinch/lumberjack.v2 v2.0.0 // indirect
+	gopkg.in/square/go-jose.v2 v2.3.1 // indirect
+	gopkg.in/yaml.v2 v2.2.2
+	k8s.io/api v0.0.0
+	k8s.io/apimachinery v0.0.0
+	k8s.io/apiserver v0.0.0
+	k8s.io/client-go v0.0.0
+	k8s.io/cloud-provider v0.0.0
+	k8s.io/cri-api v0.0.0
+	k8s.io/csi-api v0.0.0-20190313123203-94ac839bf26c // indirect
+	k8s.io/gengo v0.0.0-20190327210449-e17681d19d3a // indirect
+	k8s.io/klog v0.3.3 // indirect
+	k8s.io/kube-openapi v0.0.0-20190718094010-3cf2ea392886 // indirect
+	k8s.io/kubernetes v1.15.1
+	k8s.io/node-api v0.0.0-20190717025432-9e6fdeee55cc // indirect
+	k8s.io/utils v0.0.0-20190712204705-3dccf664f023
+	sigs.k8s.io/structured-merge-diff v0.0.0-20190628201129-059502f64143 // indirect
+)
+
+replace (
+	github.com/Sirupsen/logrus v1.0.5 => github.com/sirupsen/logrus v1.0.5
+	github.com/Sirupsen/logrus v1.3.0 => github.com/Sirupsen/logrus v1.0.6
+	github.com/Sirupsen/logrus v1.4.0 => github.com/sirupsen/logrus v1.0.6
+	k8s.io/api v0.0.0 => k8s.io/api v0.0.0-20190720062849-3043179095b6
+	k8s.io/apiextensions-apiserver v0.0.0 => k8s.io/apiextensions-apiserver v0.0.0-20190718185103-d1ef975d28ce // indirect
+	k8s.io/apimachinery v0.0.0 => k8s.io/apimachinery v0.0.0-20190612165923-1799e75a0719
+	k8s.io/apiserver v0.0.0 => k8s.io/apiserver v0.0.0-20190718184206-a1aa83af71a7
+	k8s.io/cli-runtime v0.0.0 => k8s.io/cli-runtime v0.0.0-20190718185405-0ce9869d0015
+	k8s.io/client-go v0.0.0 => k8s.io/client-go v0.0.0-20190718183610-8e956561bbf5 // indirect
+	k8s.io/cloud-provider v0.0.0 => k8s.io/cloud-provider v0.0.0-20190718190308-f8e43aa19282 // indirect
+	k8s.io/cluster-bootstrap v0.0.0 => k8s.io/cluster-bootstrap v0.0.0-20190718190146-f7b0473036f9
+	k8s.io/code-generator v0.0.0 => k8s.io/code-generator v0.0.0-20190612165923-18da4a14b22b
+	k8s.io/component-base v0.0.0 => k8s.io/component-base v0.0.0-20190718183727-0ececfbe9772
+	k8s.io/cri-api v0.0.0 => k8s.io/cri-api v0.0.0-20190531030430-6117653b35f1
+	k8s.io/csi-api v0.0.0 => k8s.io/csi-api v0.0.0-20190313123203-94ac839bf26c // indirect
+	k8s.io/csi-translation-lib v0.0.0 => k8s.io/csi-translation-lib v0.0.0-20190718190424-bef8d46b95de
+	k8s.io/gengo v0.0.0 => k8s.io/gengo v0.0.0-20190327210449-e17681d19d3a // indirect
+	k8s.io/heapster => k8s.io/heapster v1.2.0-beta.1 // indirect
+	k8s.io/klog => k8s.io/klog v0.3.3 // indirect
+	k8s.io/kube-aggregator v0.0.0 => k8s.io/kube-aggregator v0.0.0-20190718184434-a064d4d1ed7a
+	k8s.io/kube-controller-manager v0.0.0 => k8s.io/kube-controller-manager v0.0.0-20190718190030-ea930fedc880
+	k8s.io/kube-openapi v0.0.0 => k8s.io/kube-openapi v0.0.0-20190718094010-3cf2ea392886 // indirect
+	k8s.io/kube-proxy v0.0.0 => k8s.io/kube-proxy v0.0.0-20190718185641-5233cb7cb41e
+	k8s.io/kube-scheduler v0.0.0 => k8s.io/kube-scheduler v0.0.0-20190718185913-d5429d807831
+	k8s.io/kubelet v0.0.0 => k8s.io/kubelet v0.0.0-20190718185757-9b45f80d5747
```

Comment:
cant we have release sha's for dependencies ?

Reply:
go mod can use two different types of dependency versions:
1. vx.y.z where x,y,z, is an integer value
2. v0.0.0-timestamp-<first 12 characters of the commit hash>

But I think I didn't get your point can you explain me what do you mean with sha's (